### PR TITLE
DynamicAutomationPeer Refactor

### PR DESCRIPTION
**Pull request by jonthysell**
* ViewPanel, ViewControl are now closer to proper CppWinRT types
* Moved RN accessibility property storage into attached properties via DynamicAutomationProperties
* DynamicAutomationPeer now uses DynamicAutomationProperties for its logic

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/2678)

### CODE REVIEW

File: vnext/ReactUWP/Views/DynamicAutomationPeer.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  // have a way to specify the linkage between the two
+
+  // Returning nothing until that linkage exists
+
+  return nullptr;
+}
+
+void DynamicAutomationPeer::AddToSelection() const
+{
+
```
Comment (chrisglein) [inline vnext/ReactUWP/Views/DynamicAutomationPeer.cpp]: Are these not yet implemented?

File: vnext/ReactUWP/Views/DynamicAutomationPeer.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+winrt::react::uwp::AccessibilityRoles DynamicAutomationPeer::GetAccessibilityRole() const
+{
+  try
+  {
+    if (auto owner = Owner())
+    {
+      return DynamicAutomationProperties::GetAccessibilityRole(owner);
+    }
+  }
+  catch (...) {}
```
Comment (chrisglein) [inline vnext/ReactUWP/Views/DynamicAutomationPeer.cpp]: What sorts of exceptions show up here?

Comment (chrisglein) [reply vnext/ReactUWP/Views/DynamicAutomationPeer.cpp]: Same question for GetAccessibilityState and GetAccessibilityInvokeEventHandler

File: vnext/ReactUWP/Views/ViewControl.h [hunk 3]
```diff
@@ -4,101 +4,35 @@
...
 //
-struct ViewControl : winrt::Windows::UI::Xaml::Controls::ContentControlT<ViewControl>
+struct ViewControl : ViewControlT<ViewControl>
 {
-  using Super = winrt::Windows::UI::Xaml::Controls::ContentControlT<ViewControl>;
-private:
+  using Super = ViewControlT<ViewControl>;
+public:
   // Constructors
   ViewControl();
```
Comment (ahimberg) [inline vnext/ReactUWP/Views/ViewControl.h]: does the ctor need to be public or just need a different friend helper?  We had run into some accidental memory leaks before of constructing things with operator new instead of the make helper - just want to be confident we can spot those.

File: vnext/ReactUWP/Views/FrameworkElementViewManager.cpp [hunk 4]
```diff
@@ -304,6 +310,62 @@ void FrameworkElementViewManager::UpdateProperties(ShadowNodeBase* nodeToUpdate,
         }
         AnnounceLiveRegionChangedIfNeeded(element);
       }
+      else if (propertyName == "accessibilityRole")
+      {
+        if (propertyValue.isString())
+        {
+          const std::string& role = propertyValue.getString();
+          if (role == "none")
+            DynamicAutomationProperties::SetAccessibilityRole(element, winrt::react::uwp::AccessibilityRoles::None);
```
Comment (ahimberg) [inline vnext/ReactUWP/Views/FrameworkElementViewManager.cpp]: So only <View>  viewpanel/viewcontrol create the dynamicautomationpeer, if you set these properties on some other element type (an image or something) - whta happens?  the property is set but nothing is looking for it until we update everything to create custom automation peers?

File: vnext/ReactUWP/Views/ViewPanel.h [hunk 5]
```diff
@@ -50,40 +48,58 @@ struct ViewPanel : winrt::Windows::UI::Xaml::Controls::PanelT<ViewPanel>
...
   void ClipChildren(bool value);
 
-public:
   // ViewPanel Properties
-  static winrt::Windows::UI::Xaml::DependencyProperty BorderBrushProperty();
   static winrt::Windows::UI::Xaml::DependencyProperty BorderThicknessProperty();
+  static winrt::Windows::UI::Xaml::DependencyProperty BorderBrushProperty();
   static winrt::Windows::UI::Xaml::DependencyProperty CornerRadiusProperty();
-  static winrt::Windows::UI::Xaml::DependencyProperty BackgroundProperty();
   static winrt::Windows::UI::Xaml::DependencyProperty ClipChildrenProperty();
```
Comment (randy-flynn) [inline vnext/ReactUWP/Views/ViewPanel.h]: Why is this one removed? I think the new code will mess up the desired behavior. This property is separate from Panel::BackgroundProperty because it can be applied to either an outer Border element or to the Panel itself. This is a decision made in FinalizeProperties starting on line 332.

**Comment by randy-flynn** [comment]
>   }

Can you remove this duplicate code? You didn't add it of course. It was likely an accidental check-in by me.

---
Refers to: vnext/ReactUWP/Views/ViewPanel.cpp:353 in 23e2864. [](commit_id = 23e2864ace85fc5ad6ad549729b4dfc4f399d138, deletion_comment = False)

File: vnext/ReactUWP/Views/ViewViewManager.h [hunk 6]
```diff
@@ -3,12 +3,12 @@
 
 #pragma once
 
+#include <Views/ViewPanel.h>
 #include <Views/FrameworkElementViewManager.h>
```
Comment (randy-flynn) [inline vnext/ReactUWP/Views/ViewViewManager.h]: Why is this needed now? It's still just a reference, right?

File: vnext/ReactUWP/Views/DynamicAutomationProperties.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+#pragma once
+
+#include <winrt/Windows.UI.Xaml.h>
+
+#include "cppwinrt/DynamicAutomationProperties.g.h"
+
+namespace winrt::react::uwp::implementation
+{
+
+struct DynamicAutomationProperties : DynamicAutomationPropertiesT<DynamicAutomationProperties>
```
Comment (chrisglein) [inline vnext/ReactUWP/Views/DynamicAutomationProperties.h]: Not just specific to this type, but I feel like a lot of the new types you've added here could use some top level "this is what this class is for" type comment blocks.

File: vnext/ReactUWP/Views/FrameworkElementViewManager.cpp [hunk 8]
```diff
@@ -304,6 +310,62 @@ void FrameworkElementViewManager::UpdateProperties(ShadowNodeBase* nodeToUpdate,
         }
         AnnounceLiveRegionChangedIfNeeded(element);
       }
+      else if (propertyName == "accessibilityRole")
+      {
+        if (propertyValue.isString())
+        {
+          const std::string& role = propertyValue.getString();
+          if (role == "none")
```
Comment (chrisglein) [inline vnext/ReactUWP/Views/FrameworkElementViewManager.cpp]: Looks like this name to enum could be described with a map. Not sure how often this is _really_ going to get run so don't know how much the hash will help. But... could be cleaner?

File: vnext/ReactUWP/Views/ViewControl.h (hunk 3 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/ViewControl.h]: I'm trying to move us closer in alignment with a "standard" cppwinrt. If we're going to write in cppwinrt, then we need to get used to the paradigm. However, I'm still trying to determine why we I keep hitting com exceptions when using the `auto control = winrt::react::uwp::ViewControl()`.

File: vnext/ReactUWP/Views/FrameworkElementViewManager.cpp (hunk 4 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/FrameworkElementViewManager.cpp]: Yes. We can update other controls to use DynamicAutomationPeer (unfortunately via subclassing, but that's the only method XAML gives us) but now third party developers creating custom view managers can take advantage of our accessibility work.

File: vnext/ReactUWP/Views/ViewPanel.h (hunk 5 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/ViewPanel.h]: Working on a fix, will rename the property so it doesn't conflict with Panel::BackgroundProperty.

File: vnext/ReactUWP/Views/ViewControl.h (hunk 3 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/ViewControl.h]: I've figure it out - our Frankenstein cppwinrt implementation and dll defs mean our ActivateFactory methods are out of whack at the moment.

File: vnext/ReactUWP/Views/ViewPanel.h (hunk 5 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/ViewPanel.h]: REsolved.

File: vnext/ReactUWP/Views/DynamicAutomationPeer.cpp [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    return winrt::AutomationControlType::Text;
+  case winrt::react::uwp::AccessibilityRoles::Adjustable:
+    return winrt::AutomationControlType::Slider;
+  case winrt::react::uwp::AccessibilityRoles::Search:
+  case winrt::react::uwp::AccessibilityRoles::Unknown:
+  default:
+    return winrt::AutomationControlType::Group;
+  }
+
+  return Super::GetAutomationControlTypeCore();
```
[the remark this answers is not in the document]
Comment (jonthysell) [reply vnext/ReactUWP/Views/DynamicAutomationPeer.cpp]: Yes

File: vnext/ReactUWP/Views/DynamicAutomationPeer.cpp (hunk 2 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/DynamicAutomationPeer.cpp]: We're not reference tracking properly. So when UIA grabs the peer, there's a case that the Owner will have been released. We're using an old cppwinrt - I think it'll be fixed when we use a newer version, or we may need to take the reference tracking stuff that mux uses to implement the correct lifetime behavior. Neither are small tasks.

Comment (jonthysell) [reply vnext/ReactUWP/Views/DynamicAutomationPeer.cpp]: In the meantime, this prevents UIA tools from crashing the app.

File: vnext/ReactUWP/Views/FrameworkElementViewManager.cpp (hunk 8 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/FrameworkElementViewManager.cpp]: Yeah, we have a lot of cleanup work in this codebase for all of these string compares. But it might be something that waits for fabric, when we have to replace the viewmanagers anyway.

File: vnext/ReactUWP/Views/DynamicAutomationProperties.h (hunk 7 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/DynamicAutomationProperties.h]: REsolved.

File: vnext/ReactUWP/Views/DynamicAutomationPeer.cpp (hunk 1 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/DynamicAutomationPeer.cpp]: Right now RN does not have "selection" events, so this is a no-op

File: vnext/ReactUWP/Views/ViewViewManager.h (hunk 6 above)
Comment (jonthysell) [reply vnext/ReactUWP/Views/ViewViewManager.h]: TryUpdateView needs it.

**Comment by jonthysell** [comment]
> > 
> 
> Can you remove this duplicate code? You didn't add it of course. It was likely an ac

@randy-flynn I can't seem to be able to see what code you're referring to here.

**Comment by randy-flynn** [comment]
>   }

Seems like the CodeFlow/GitHub synchronization failed me h ere. Opening a new comment.

---
In reply to: [505673059](https://github.com/microsoft/react-native-windows/pull/2678#issuecomment-505673059) [](ancestors = 505673059)

---
Refers to: vnext/ReactUWP/Views/ViewPanel.cpp:353 in 23e2864. [](commit_id = 23e2864ace85fc5ad6ad549729b4dfc4f399d138, deletion_comment = False)

File: vnext/ReactUWP/Views/DynamicAutomationPeer.cpp (hunk 2 above)
Comment (chrisglein) [reply vnext/ReactUWP/Views/DynamicAutomationPeer.cpp]: Can you leave a comment in these spots to that effect? Catch alls like this tend to stick once added because no one has the courage to remove. It helps if there's a breadcrumb.

**Comment by jonthysell** [comment]
> > }
> 
> Seems like the CodeFlow/GitHub synchronization failed me h ere. Opening a new comment.
> 
> In reply to: [505673059](https://github.com/microsoft/react-native-windows/pull/2678#issuecomment-505673059) [](ancestors = 505673059)
> 
> Refers to: vnext/ReactUWP/Views/ViewPanel.cpp:353 in [23e2864](https://github.com/microsoft/react-native-windows/commit/23e2864ace85fc5ad6ad549729b4dfc4f399d138). [](commit_id = [23e2864](https://github.com/microsoft/react-native-windows/commit/23e2864ace85fc5ad6ad549729b4dfc4f399d138), deletion_comment = False)

Ok, I've removed the duplicated code.

I've found the CodeFlow integration to be weird at best.

**Comment by jonthysell** [comment]
Whoops, I did not mean close the PR.

File: vnext/ReactUWP/Views/ViewViewManager.cpp [hunk 10]
```diff
@@ -316,16 +296,18 @@ void ViewViewManager::UpdateProperties(ShadowNodeBase* nodeToUpdate, const folly
       const std::string& propertyName = pair.first.getString();
       const folly::dynamic& propertyValue = pair.second;
 
-      if (TryUpdateBackgroundBrush(*pPanel, propertyName, propertyValue))
+      if (propertyName == "backgroundColor")
```
Comment (randy-flynn) [inline vnext/ReactUWP/Views/ViewViewManager.cpp]: It would be nice to update the property helper to handle this case, though that's definitely not needed for this check-in.
